File: pl/portscan.pl
#!/usr/bin/perl -w
use lib '/tmp/mytools/pl/lib';
use threads; 
use Socket;
use Net::IP;
use Thread::Queue;
use Thread::Semaphore; 

# flush the print buffer immediately
$| = 1;
my $q = Thread::Queue-> new(); 

my $mutex = Thread::Semaphore->new( 1 );   #mutex

sub scan{
		local($target,$port)=@_;
	    $ip_a= sockaddr_in($port, inet_aton($target)); 
        $proto=getprotobyname('tcp');  
        socket(SOCK, PF_INET, SOCK_STREAM,$proto);
        if(connect(SOCK,$ip_a))
		{ 
             print "[#] host:$target:$port is Open! [#]\n";
   }
          close(SOCK);
}

sub worker { 
		my($target)=@_;
		while(1) { 
				$mutex ->down();
				my $r = $q->dequeue_nb();
				$mutex ->up();
				if ($r){
						&scan($target,$r)
				}
				else{
						last;
				} 
   } 
}

my @portlist;
if (index($ARGV[1],"-")>0){
		@rport=split(/-/,$ARGV[1]);
		for($port=$rport[0];$port<=$rport[1];$port++){
				push @portlist,$port;
		}
}
elsif (index($ARGV[1],",")>0){
		@portlist=split(/,/,$ARGV[1]);
}
else{
		push @portlist,$ARGV[1];
}

foreach $port (@portlist){
		$q->enqueue($port);
}


my $nb_process = 100;
if (@ARGV == 3){
		$nb_process = $ARGV[2];}
my $tn=0;
my @Threads;
while ($tn < $nb_process) {
 		my $thread = threads->create(\&worker, $ARGV[0]);
		push (@Threads, $thread);
		$tn = $tn+1;
		}
foreach my $thr (@Threads)
{
$thr->join;
}



#my $ip = new Net::IP ($ARGV[0]) or die (Net::IP::Error());
#my @iplist;
#do {
#   push @iplist, $ip->ip();
#} while (++$ip);
#print @iplist[10]
#&scan($ARGV[0],$ARGV[1]);

File: py/portscan.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#*.py ip 1-1024 200
import sys
import socket
import threading
from Queue import Queue
def scan(host,port):
		s = socket.socket()
		s.settimeout(1)
		if s.connect_ex((host, port)) == 0:
				#print port, 'open'
				counting_open.append(port)
		s.close()
 
def worker(host):
		while not q.empty():
				mutex.acquire(3)
				port = q.get()
				mutex.release()
				try:
						scan(host,int(port))
				except:
						print 1
				finally:
						q.task_done()
 
if __name__ == '__main__':
		host = sys.argv[1]
		portlist = []
		mutex = threading.Lock()
		if sys.argv[2].find('-')>=0:
				portstrs = sys.argv[2].split('-')
				start_port = int(portstrs[0])
				end_port = int(portstrs[1])
				for port in range(start_port, end_port):
						portlist.append(port)
		elif sys.argv[2].find(',')>=0:
				portlist = sys.argv[2].split(',')
		else:
				portlist.append(sys.argv[2])
		counting_open = []
		q = Queue()
		threads = []
		num_procs = 500
		if len(sys.argv) ==4:
				num_procs = int(sys.argv[3])
		map(q.put,portlist)
		for i in xrange(num_procs):
				c = threading.Thread(target=worker,args=(host,))
				c.setDaemon(True)
				threads.append(c)
		map(lambda x:x.start(),threads)
		q.join()
		print host+" Opened ports:"
		print counting_open
